### ISSUE TITLE
Minor but in autodoc.py refactor: Use get_attr to get '__dict__'

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#107]
One should use the general get_attr method to allow for a customized version of **dict** to be returned by get_object_member().

This is required for example with zope interfaces because the Interface metaclass removes all attributes from **dict** and one needs to use names() to get them.

Here is a patch against is attached revision 9f5725038f9d.  The patch also contains two minor documentation changes, including a comment related to issue #106.

```
diff -r 9f5725038f9d sphinx/ext/autodoc.py
--- a/sphinx/ext/autodoc.py Thu Feb 19 16:15:59 2009 +0100
+++ b/sphinx/ext/autodoc.py Thu Feb 19 09:36:10 2009 -0700
@@ -214,6 +214,11 @@
     The *special_attrgetters* attribute is used to customize ``getattr()``
     calls that the Documenter makes; its entries are of the form
     ``type: getattr_function``.
+
+    .. note:: When importing an object, all items along the import
+       chain are accessed using the descendant's
+       *special_attrgetters*, thus this dictionary should include all
+       necessary functions for accessing attributes of the parents.
     """
     #: name by which the directive is called (auto...) and the default
     #: generated directive name
@@ -454,7 +459,8 @@

     def get_object_members(self, want_all):
         """
-        Return (membername, member) pairs of the members of *self.object*.
+        Return `(members_check_module, members)` where `members` is a
+        list of `(membername, member)` pairs of the members of *self.object*.

         If *want_all* is True, return all members.  Else, only return those
         members given by *self.options.members* (which may also be none).
@@ -479,8 +485,9 @@
             # __dict__ contains only the members directly defined in
             # the class (but get them via getattr anyway, to e.g. get
             # unbound method objects instead of function objects)
-            return False, sorted([(mname, self.get_attr(self.object, mname))
-                                  for mname in self.object.__dict__])
+            return False, sorted([
+                (mname, self.get_attr(self.object, mname))
+                for mname in self.get_attr(self.object, '__dict__')])

     def filter_members(self, members, want_all):
         """
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/107
- Originally reported by: Michael Forbes
- Originally created at: 2009-02-19T17:38:41.847

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-02-19 21:55:34+00:00_

Fixed in <<cset 1d128f57fa99de52b0a83e4357bb467cea867803>>.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-08-25 12:19:55+00:00_

Removing milestone: 0.6 (automated comment)
